libs/ast_generic/AST_utils.ml

```diff
@@ -40,26 +40,57 @@ module Node_ID = Gensym.MkId ()
    so as to select structural or referential equality.
 *)
 
-type busy_with_equal = Not_busy | Structural_equal | Referential_equal
+type busy_with_equal =
+  | Not_busy
+  | Structural_equal
+  | Referential_equal
+  | Syntactic_equal
 
 (* global state! managed by the with_equal_* functions *)
 let busy_with_equal = ref Not_busy
 
+let equal_id_info equal a b =
+  match !busy_with_equal with
+  | Not_busy -> failwith "Call AST_utils.with_xxx_equal to avoid this error."
+  | Syntactic_equal -> true
```

Comment:
need a comment explaining why for syntatic_equal we don't care about id_info, and in which context we need that (I guess for join-mode, but could be nice to repeat it here).